src/main/java/com/maxmind/db/Decoder.java

```diff
@@ -272,42 +288,50 @@ private static boolean decodeBoolean(int size)
         Object array = Array.newInstance(elementClass, size);
 
         for (int i = 0; i < size; i++) {
-            Object e = this.decode(elementClass);
+            Object e = this.decode(elementClass, null);
             Array.set(array, i, e);
         }
 
         return array;
     }
 
-    private <T> Object decodeMap(int size, Class<T> cls)
-            throws IOException,
-                   InstantiationException,
-                   IllegalAccessException,
-                   InvocationTargetException {
-        // We look for Object.class because if we're decoding a nested map into
-        // a Map, the second level is an Object.class. See Object.class in
-        // decodeMapIntoMap().
-        if (cls.isAssignableFrom(Map.class) || cls.equals(Object.class)) {
-            return this.decodeMapIntoMap(size);
+    private <T> Object decodeMap(
+            int size,
+            Class<T> cls,
+            java.lang.reflect.Type genericType
+    ) throws IOException,
+             InstantiationException,
+             IllegalAccessException,
+             InvocationTargetException {
+        if (cls.isAssignableFrom(Map.class)) {
+            return this.decodeMapIntoMap(size, genericType);
         }
 
         return this.decodeMapIntoObject(size, cls);
     }
 
-    private HashMap<String, Object> decodeMapIntoMap(int size)
-            throws IOException,
-                   InstantiationException,
-                   IllegalAccessException,
-                   InvocationTargetException {
+    private HashMap<String, Object> decodeMapIntoMap(
+            int size,
+            java.lang.reflect.Type genericType
+    ) throws IOException,
+             InstantiationException,
+             IllegalAccessException,
+             InvocationTargetException {
         HashMap<String, Object> map = new HashMap<>();
 
-        for (int i = 0; i < size; i++) {
-            String key = (String) this.decode(String.class);
-
-            // Ideally we'd know the value's type to pass in, but I am having
-            // trouble retrieving it. I am not sure it is possible.
-            Object value = this.decode(Object.class);
+        Class<?> valueClass = Object.class;
+        if (genericType instanceof ParameterizedType) {
+            ParameterizedType pType = (ParameterizedType) genericType;
+            java.lang.reflect.Type[] actualTypes
+                = pType.getActualTypeArguments();
+            if (actualTypes.length == 2) {
+                valueClass = (Class<?>) actualTypes[1];
```

Comment:
Maybe we should do the same for the key class or at least check that it is a `String.class`?

Reply:
I was thinking that since maps always have to be strings, we might not need to? Oh, I can add a check at least.

Reply:
Map keys that is.